### ISSUE TITLE
Build reconfiguration should consistently honor env variables

### ISSUE THREAD

**Issue by xuhdev** [pytorch/pytorch#23512]
Per discussion here: https://github.com/pytorch/pytorch/pull/23323#issuecomment-515168182

We currently have

(plainly == no additional build options passed in to setup.py)

for clean tree:

- configuration only: `--cmake-only`
- configuration + build: invoke setup.py plainly

For rebuild:

- reconfiguration without rebuild: `--cmake-only --cmake` (unreliable), or edit CMakeCache.txt and run cmake directly
- rebuild without reconfiguration: invoke setup.py plainly
- reconfiguration + rebuild: `--cmake` (unreliable), or edit CMakeCache.txt and invoke setup.py plainly

We need to make the two "unreliable" spots "reliable".

**Comment by colesbury** [comment]
Yes please. @xuhdev do you plan to work on this?

**Comment by xuhdev** [comment]
@colesbury Yes, but there are more cleanup to do before this can be done (i.e., lots of detections should be moved to CMake, see #21702). This is current set as a placeholder so we can bear this in mind while continuing build cleanup.
